weaviate/collections/batch/grpc_batch_objects.py

```diff
@@ -57,7 +57,7 @@ def pack_vector(vector: Any) -> bytes:
                 collection=obj.collection,
                 vector_bytes=(
                     pack_vector(obj.vector)
-                    if obj.vector is not None and isinstance(obj.vector, list)
+                    if obj.vector is not None and not isinstance(obj.vector, dict)
```

Comment:
I think handling of vector input should be at the input level, eg in insert_many and batch.add_objects and not here

Reply:
Could you please elaborate what you mean by this and some hints how to accomplish?

I was staring at the code and don't see how I could push the validation upstream, since `obj.vector` can be both `None` and a `dict`. Couple of lines below where `vector` is constructed: `if obj.vector is not None and isinstance(obj.vector, dict)`, which means either vector_bytes or vector will be not-None, but not both at the same time.
For readability, we could move these two next to each other - just cosmetics.

Note: the only thins I would do is to move `pack_vector` outside as is `_pack_named_vectors`, and rename the latter to `_pack_named_vector`. But that is again just cosmetics :)

Reply:
I think the correct place would be 'weaviate/collections/data/data.py::insert_many` and then the cleanup should be here `vector=obj.vector`. But keep in mind that vectors can be lists and dictionaries (for named vectors) and both should be supported

Reply:
Hi @dirkkul, thanks for coming back. However, I still don't see how this could be achieved without a major refactoring.

There are 3 possibilities in the current program flow, starting from `data/_DataCollectionAsync.insert_many()`:

| type(obj)  | type(obj.vector) | _BatchObject.vector | vector_bytes | vectors |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| !DataObject  | n/a  | None | None | None |
| DataObject  | !dict  | obj.vector | pack_vector(...) | None |
| DataObject  | dict  | obj.vector | None | _pack_named_vectors(...) |

So move one needs to move some of the logic from `__grpc_objects()` into `insert_many()`, e.g. creating a list of tuples containing (_BatchObject, vector_bytes, vectors). Is this what you are suggesting or am I on the wrong path please?